# KNN - Classificação Câncer de Mama

In [0]:
import pandas as pd
import numpy as np
import math
import operator

#Carrega o dataset
data = pd.read_csv("train_dataset.csv")
data.head()


1183516  3  1  1.1  1.2  2  1.3  1.4  1.5  1.6  2.1
0  1183911  2  1    1    1  2    1    1    1    1    2
1  1183983  9  5    5    4  4    5    4    3    3    4
2  1184184  1  1    1    1  2    5    1    1    1    2
3  1184241  2  1    1    1  2    1    2    1    1    2
4  1184840  1  1    3    1  2    4    2    1    1    2

In [0]:
# Define a função que calcula a distância Euclidiana entre dois pontos(atributos).
def euclideanDistance(data1, data2, length):
    distance = 0
    for x in range(length):
        distance += np.square(data1[x] - data2[x])
    return np.sqrt(distance)

# Define o modelo KNN
def knn(dataset, testRow, k): 
    
    distances = {}
    sort = {}
    length = 9
   # print("len test:", length);
    
    
    # Calculo da distância Euclidiana entre cada linha dos dados de treinamento e teste
    for x in range(len(dataset)):
         
        dist = euclideanDistance(testRow, dataset.iloc[x], length)
        
        distances[x] = dist
        
    # Ordena o vetor em ordem crescente
    sorted_d = sorted(distances.items(), key=operator.itemgetter(1))
    
    
    neighbors = []
    
    # Pega o top k vizinhos
    
    for x in range(k):
        neighbors.append(sorted_d[x][0])
    
    
    #votos de cada classe
    classVotes = {}
    
    
    # Cálculo da maior frequência que uma classe aparece nos vizinhos
    for x in range(len(neighbors)):
        response = dataset.iloc[neighbors[x]][-1]
        
        if response in classVotes:
            classVotes[response] += 1
        else:
            classVotes[response] = 1
    

    #Ordena do maior para o menor(ranking de quantos vizinhos próximos existem de cada classe)
    sortedVotes = sorted(classVotes.items(), key=operator.itemgetter(1), reverse=True)
    return(sortedVotes[0][0], neighbors)
   



In [0]:
data_test = pd.read_csv("test_dataset.csv")
#data_test_label = pd.read_csv("test_dataset.csv")
#data_test.head()


In [0]:
#tranformando o dataset das labels corretas
data_test_label = data_test.drop(['5', '1', '2', '3', '4', '6', '7', '8', '9', '10'], axis= 1)
lista_label = data_test_label.values.tolist()
print(lista_label)
#transformando o dataset para teste
data_test = data_test.drop([ '1', '11'], axis= 1)
lista_test = data_test.values.tolist()
print(lista_test)
#data_test.head()

[[2], [2], [2], [2], [2], [4], [2], [2], [2], [2], [2], [2], [4], [2], [4], [4], [2], [2], [4], [2], [4], [4], [2], [4], [2], [4], [2], [2], [2], [2], [2], [2], [4], [2], [2], [2], [4], [2], [4], [4], [2], [4], [4], [4], [4], [2], [4], [2], [2], [4], [4], [4], [4], [4], [4], [4], [4], [4], [4], [4], [4], [2], [4], [4], [2], [4], [2], [4], [4], [2], [2], [4], [2], [4], [4], [2], [2], [2], [2], [2], [2], [2], [2], [2], [4], [4], [4], [4], [2], [2], [2], [2], [2], [2], [2], [2], [2], [2], [4], [4], [4], [4], [2], [4], [4], [4], [4], [4], [2], [4], [2], [4], [4], [4], [2], [2], [2], [4], [2], [2], [2], [2], [4], [4], [4], [2], [4], [2], [4], [2], [2], [2], [4], [2], [2], [2], [2], [2], [2], [2]]
[[5, 1, 1, 1, 2, 1, 3, 1, 1], [5, 4, 4, 5, 7, 10, 3, 2, 1], [3, 1, 1, 1, 2, 2, 3, 1, 1], [6, 8, 8, 1, 3, 4, 3, 7, 1], [4, 1, 1, 3, 2, 1, 3, 1, 1], [8, 10, 10, 8, 7, 10, 9, 7, 1], [1, 1, 1, 1, 2, 10, 3, 1, 1], [2, 1, 2, 1, 2, 1, 3, 1, 1], [2, 1, 1, 1, 2, 1, 1, 1, 5], [4, 2, 1, 1, 2, 1, 2, 1, 1], [1,

In [0]:

#Testando o algoritmo


#Setando o número de vizinhos
k = 1

total1 = 0

for x in range(len(lista_test)):
    # Executando o modelo KNN para cada linha do dataset de test
    predict = lista_label[x]
    
    result,neigh = knn(data, lista_test[x], k)
   # print(result)
    #print(neigh)
    #print(predict)
    if (result == predict):
        total1+=1
    
print("Acurácia foi de: ", 100*total1/len(lista_test))
#testRow1 = [[4,8,8,5,4,5,10,4,1]] # resposta certa é 4
#test = pd.DataFrame(testRow1)



#[[1,24,5,6,7,8,9,0,0]]




Acurácia foi de:  55.714285714285715


In [0]:
# Testando com 3 vizinhos 
k = 3 

total2 = 0

for x in range(len(lista_test)):
    # Executando o modelo KNN para cada linha do dataset de test
    predict = lista_label[x]
    
    result,neigh = knn(data, lista_test[x], k)
   # print(result)
    #print(neigh)
    #print(predict)
    if (result == predict):
        total2+=1
    
print("Acurácia foi de: ", 100*total2/len(lista_test))
#testRow1 = [[4,8,8,5,4,5,10,4,1]] # resposta certa é 4
#test = pd.DataFrame(testRow1)


Acurácia foi de:  44.285714285714285


In [0]:
# Testando com 5 vizinhos 
k = 5
 
total3 = 0

for x in range(len(lista_test)):
    # Executando o modelo KNN para cada linha do dataset de test
    predict = lista_label[x]
    
    result,neigh = knn(data, lista_test[x], k)
   # print(result)
    #print(neigh)
    #print(predict)
    if (result == predict):
        total3+=1
    
print("Acurácia foi de: ", 100*total3/len(lista_test))
#testRow1 = [[4,8,8,5,4,5,10,4,1]] # resposta certa é 4
#test = pd.DataFrame(testRow1)


Acurácia foi de:  44.285714285714285


# Relatório

---

Como o KNN usa os vizinhos mais próximos para tomar uma decisão e classificar um novo problema, é preciso levar em conta o método escolhido que mede a distância e principalmente o valor do "k". O valor do "k" define quantos vizinhos serão escolhidos para tomar a decisão, sendo que, a classe que tiver maior quantidade de vizinhos vence. Então cosneguimos testar e ver que se o "k" é muito pequeno, o algoritmo nem sempre vai ser tão preciso em alguns casos. Observamos também que não há uma grande diferença entre os k's depois que ele converge.(No caso, os resultados com k=3 eram bem parecidos com os de k=5.

---

Acurácia k=1 : 55.71%     

Acurácia k=3 : 44.28%

Acurácia k=3 : 44.28%

---

Alunos: Alexandre Bender, Victor Grovermann, Moniele Santos, Vinicius Borges, Rafael Ferreira